tiledb/sm/storage_manager/consolidator.cc

```diff
@@ -64,14 +73,227 @@ Status Consolidator::consolidate(
+  for (int i = 0; i <= (int)start - 1; ++i) {
+    if (utils::geometry::overlap(
+            union_non_empty_domains,
+            (T*)fragments[i].non_empty_domain_,
+            dim_num))
+      return false;
+  }
+
+  // Check consolidation amplification factor
+  auto union_cell_num =
+      utils::geometry::cell_num<T>(union_non_empty_domains, dim_num);
+  uint64_t sum_cell_num = 0;
+  for (size_t i = start; i <= end; ++i) {
+    sum_cell_num += utils::geometry::cell_num<T>(
+        (T*)fragments[i].non_empty_domain_, dim_num);
+  }
+  return (double(union_cell_num) / sum_cell_num) <= config_.amplification_;
+}
+
+Status Consolidator::consolidate(
+    const ArraySchema* array_schema,
+    EncryptionType encryption_type,
+    const void* encryption_key,
+    uint32_t key_length) {
+  switch (array_schema->coords_type()) {
+    case Datatype::INT32:
+      return consolidate<int>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::INT64:
+      return consolidate<int64_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::INT8:
+      return consolidate<int8_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::UINT8:
+      return consolidate<uint8_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::INT16:
+      return consolidate<int16_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::UINT16:
+      return consolidate<uint16_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::UINT32:
+      return consolidate<uint32_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::UINT64:
+      return consolidate<uint64_t>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::FLOAT32:
+      return consolidate<float>(
+          array_schema, encryption_type, encryption_key, key_length);
+    case Datatype::FLOAT64:
+      return consolidate<double>(
+          array_schema, encryption_type, encryption_key, key_length);
+    default:
+      return LOG_STATUS(
+          Status::ConsolidatorError("Cannot consolidate; Invalid domain type"));
+  }
+
+  return Status::Ok();
+}
+
+template <class T>
+Status Consolidator::consolidate(
+    const ArraySchema* array_schema,
+    EncryptionType encryption_type,
+    const void* encryption_key,
+    uint32_t key_length) {
+  std::vector<FragmentInfo> to_consolidate;
+  auto timestamp = utils::time::timestamp_now_ms();
+  auto array_uri = array_schema->array_uri();
+  EncryptionKey enc_key;
+  RETURN_NOT_OK(enc_key.set_key(encryption_type, encryption_key, key_length));
+  auto domain_size = 2 * array_schema->coords_size();
+
+  // Allocate memory for union of non-empty domains of fragments to consolidate
+  auto union_non_empty_domains = (T*)std::malloc(domain_size);
```

Comment:
Can we start using either stack allocations or `std::unique_ptr<uint8_t>` for these ad-hoc allocations? We shouldn't be introducing more malloc/frees than necessary.